Add systemd-nspawn connection driver

tszym opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
connection plugin

##### ANSIBLE VERSION
```
ansible 2.3.0
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

This PR follows #14334 from @larsks as it seems like he is not following his PR anymore (so all credits for this work to him) and some people are willing to benefit from this feature.
I adapted the patch according to @bcoca comments and did the rebase to have a patch that could be merged.

From @larsks words:

This commit adds a connection driver built on top of systemd-nspawn.
This is similar to the existing `chroot` driver, except that nspawn
offers a variety of additional services. For example, it takes care of
automatically mounting `/proc` and `/sys` inside the chroot environment,
which will make a variety of tools work correctly that would otherwise
fail.

You can take advantage of other system-nspawn features to perform more
complicated tasks. For example, on my x86_64 system I have a Raspberry
Pi disk image mounted on `/rpi`. I can't use chroot with this because
the binaries contained in the image are for the wrong architecture.
However, I can use the systemd-nspawn `--bind` option to automatically
insert the appropriate qemu-arm binary into the container using an
inventory file like this:

```
pi ansible_host=/rpi ansible_nspawn_extra_args='--bind /usr/bin/qemu-arm --bind /lib64'
```

See http://www.freedesktop.org/software/systemd/man/systemd-nspawn.html
for more information about systemd-nspawn itself.

Please take a look at #14334 for the full history about this.

larsks committed:
add systemd-nspawn connection driver

This commit adds a connection driver built on top of systemd-nspawn.
This is similar to the existing `chroot` driver, except that nspawn
offers a variety of additional services. For example, it takes care of
automatically mounting `/proc` and `/sys` inside the chroot environment,
which will make a variety of tools work correctly that would otherwise
fail.

You can take advantage of other system-nspawn features to perform more
complicated tasks.  For example, on my x86_64 system I have a Raspberry
Pi disk image mounted on `/rpi`.  I can't use `chroot` with this because
the binaries contained in the image are for the wrong architecture.
However, I can use the systemd-nspawn `--bind` option to automatically
insert the appropriate qemu-arm binary into the container using an
inventory file like this:

    pi ansible_host=/rpi ansible_nspawn_extra_args='--bind /usr/bin/qemu-arm --bind /lib64'

See http://www.freedesktop.org/software/systemd/man/systemd-nspawn.html
for more information about systemd-nspawn itself.

tszym committed:
Add a config section for systemd-nspawn driver

bcoca commented on lib/ansible/cli/__init__.py:
```diff
@@ -353,6 +353,8 @@ def base_parser(usage="", output_opts=False, runas_opts=False, meta_opts=False,
             connect_group.add_option('--ssh-extra-args', default='', dest='ssh_extra_args',
                 help="specify extra arguments to pass to ssh only (e.g. -R)")
+            connect_group.add_option('--nspawn-extra-args', default='', dest='nspawn_extra_args',
+                help="specify extra arguments to pass to systemd-nspawn only (e.g. --bind)")
```
i prefer to not have extra command line args (really want to remove `--ssh-extra` and use a `--connection-extra` instead).

tszym replied:
I will remove this option as there is a section added to `ansible.cfg` already.

tszym committed:
Remove the --nspawn-extra-args cli option

Providing extra arguments to pass to systemd-nspawn is still possible
through the `nspawn_connection` of the `ansible.cfg` file.

tszym commented:
Looks like shippable tests failed on some unrelated network timeouts.
Am I misunderstanding?

tszym commented:
@bcoca are you satisfied with the changes?

larsks commented:
Thanks @tszym for taking the time to bring my original PR up-to-date!  I haven't so much stopped following the original PR as much as I have sort of thrown in the towel w/r/t trying to submit patches to ansible.  I'm glad you found the original work useful, and I hope other people find it so also.

bcoca commented:
lgtm, added to core meeting agenda to get additional support

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+
+        self.nspawn_cmd = 'systemd-nspawn'
+
+    def _connect(self):
```
This should set self._connected = True (not really needed right now but could have ramifications later) and also use display to print a message.  For instance, this is what the docker connect plugin does:
```
    def _connect(self, port=None):
        """ Connect to the container. Nothing to do """
        super(Connection, self)._connect()
        if not self._connected:
            display.vvv(u"ESTABLISH DOCKER CONNECTION FOR USER: {0}".format(
                self.actual_user or '?'), host=self._play_context.remote_addr
            )
            self._connected = True

```

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+                if C.DEFAULT_EXECUTABLE
+                else '/bin/sh')
+
+        nspawn_args = shlex.split(self._play_context.nspawn_args)
```
Unfortunately, shlex.split() needs a bit of code to make it compatible with both python-2.6 and python3.x

On python-2.6, it only works on byte strings.  In python3.x it only works with text strings.  So you have to test for python version and then convert appropriately.  Code like the following is what I use:

``` python
from ansible.compat import six
from ansible.module_utils._text import to_bytes, to_text
[...]
nspawn_args = self._play_context.nspawn_args
if six.PY2:
    nspawn_args = shlex.split(to_bytes(nspawn_args, errors='surrogate_or_strict'))
else:
    nspawn_args = shlex.split(to_text(nspawn_args, errors='surrogate_or_strict'))

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self.chroot)
+
+        out_path = pipes.quote(self._prefix_login_path(out_path))
```
We've switched from pipes.quote()  to six.moves.shlex._quote for python3 compatibility:
``` python
 from ansible.compat.six.moves import shlex_quote
 out_path = shlex_quote(self._prefix_login_path(out_path))
```

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+                try:
+                    p = self._buffered_exec_command('dd of=%s bs=%s' % (out_path, BUFSIZE), stdin=in_file)
+                except OSError:
+                    raise AnsibleError("chroot connection requires dd command in the chroot")
```
replace "chroot connection" in messages with  "nspawn connection".

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+        super(Connection, self).fetch_file(in_path, out_path)
+        display.vvv("FETCH %s TO %s" % (in_path, out_path), host=self.chroot)
+
+        in_path = pipes.quote(self._prefix_login_path(in_path))
```
shlex_quote again

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+from __future__ import (absolute_import, division, print_function)
```
There needs to be a copyright and license header above this.  All controller-side code in ansible is GPLv3+.  You can take a look at another connection plugin (for instance, chroot.py https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/chroot.py ) for an example.

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+    from ansible.utils.display import Display
+    display = Display()
+
+BUFSIZE = 65536
```
Nowadays, we import BUFSIZE from ansible.plugins.connection.

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+from ansible.errors import AnsibleError
+from ansible.plugins.connection import ConnectionBase
+from ansible.module_utils.basic import is_executable
+from ansible.utils.unicode import to_bytes
```
The standard place to import this from is now:  from ansible.module_utils._text import to_bytes

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
```
Should import os.path here as well (it's an implementation detail that os.path works after only importing os.)

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+                '--', executable, '-c', cmd]
+
+        display.vvv("EXEC %s" % (local_cmd), host=self.chroot)
+        local_cmd = map(to_bytes, local_cmd)
```
It''s desirable to call to_bytes with errors='surrogate_or_strict' so that it raises an exception instead of replacing undecodable bytes with a "?".  Also, map() is more pythonically expressed as a list comprehension.
``` python
local_cmd = [to_bytes(i, errors='surrogate_or_strict') for i in local_cmd]
```

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+
+        chrootsh = os.path.join(self.chroot, 'bin/sh')
+        if not is_executable(chrootsh):
+            raise AnsibleError("%s does not look like a chrootable dir (/bin/sh missing)" % self.chroot)
```
I'm unsure of systemd-nspawn terminology.  Is "chroot", "chrootable", etc, in descriptions the proper words to use?

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+
+        out_path = pipes.quote(self._prefix_login_path(out_path))
+        try:
+            with open(in_path, 'rb') as in_file:
```
We should convert filenames to bytes before passing to open.  Like this:
``` python
with open(to_bytes(in_path, errors='surrogate_or_strict'), 'rb') as in_file:
```

abadger commented on lib/ansible/plugins/connection/nspawn.py:
```diff
@@ -0,0 +1,152 @@
+        except OSError:
+            raise AnsibleError("chroot connection requires dd command in the chroot")
+
+        with open(out_path, 'wb+') as out_file:
```
and convert filename to bytes again.

abadger commented:
By and large, looks good.  There's a few things that are needed and some other things that are more minor.  One other thing I'd do is use a graphical diff tool (I like meld) to compare this file to a similar connection plugin like chroot.py.  Then you can easily see where we've updated the code to handle more corner cases since this one was originally forked from chroot.

tszym committed:
Add safety checks to nspawn connection plugin

This patch adds some checks on the path that is accessed as a container,
making sure it looks like one. It implements the connection method and
add adaptations to the modern way of writing connections for Ansible.
It also rewords docs and vars to use the nspawn terminology instead of
chroot.

tszym commented:
Hi @abadger,

Thanks for this review. I updated based on your advices: the copyright bloc has been added. Python3 support modifications are done and the code is adapted according to the nspawn terminology.

I also added some checks. Hope you'll like it.

mattclay commented:
CI failure due to PEP 8 issues (PEP 8 testing was recently added to CI):

```
2017-02-03 22:44:59 ERROR: PEP 8: lib/ansible/plugins/connection/nspawn.py:89:17: E126 continuation line over-indented for hanging indent (current)
2017-02-03 22:44:59 ERROR: PEP 8: lib/ansible/plugins/connection/nspawn.py:100:17: E126 continuation line over-indented for hanging indent (current)
```

tszym commented:
Oh I see. I added some modifications to be ok with PEP 8.
However there are still E402 errors due to the `__metaclass__ = type` line. But this is the same in other connection plugins for now.

mattclay reviewed (changes requested):
@tszym Could you provide some instructions on how to test this plugin?

gundalow commented:
@tszym Hi, wondering if you have some time to finish this off? It would be good to get this into 2.3, which we will start to lock down for code freeze in just over a week

abadger commented:
Code reviewed and from that standpoint, this looks good to merge.  +1 for merging from me.

@tszym I think @mattclay would like some information so that we can get this integration tested to prevent regressions from occuring as changes occur in the code base.

abadger commented:
Okay, I saw the information in the first comment and adapted it to a chroot I have on my machine:

inventory:
```
myhost ansible_host=/var/lib/libvirt/filesystems ansible_nspawn_extra_args='--bind /usr/bin/qemu-x86_64' ansible_connection=nspawn
```
and tried to use it:
```
# ansible myhost -m setup -i inven -vvvv                                                    (09:58:17)
Using /etc/ansible/ansible.cfg as config file
Loading callback plugin minimal of type stdout, v2.0 from /srv/ansible/vanilla2/lib/ansible/plugins/callback/__init__.pyc
META: ran handlers
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/srv/ansible/vanilla2/lib/ansible/executor/task_executor.py", line 126, in run
    res = self._execute()
  File "/srv/ansible/vanilla2/lib/ansible/executor/task_executor.py", line 455, in _execute
    self._connection = self._get_connection(variables=variables, templar=templar)
  File "/srv/ansible/vanilla2/lib/ansible/executor/task_executor.py", line 712, in _get_connection
    connection = self._shared_loader_obj.connection_loader.get(conn_type, self._play_context, self._new_stdin)
  File "/srv/ansible/vanilla2/lib/ansible/plugins/__init__.py", line 374, in get
    obj = obj(*args, **kwargs)
  File "/srv/ansible/vanilla2/lib/ansible/plugins/connection/nspawn.py", line 50, in __init__
    display.vvv("NSPAWN ARGS %s" % self._play_context.nspawn_args)
AttributeError: 'PlayContext' object has no attribute 'nspawn_args'

pi | FAILED! => {
    "failed": true, 
    "msg": "Unexpected failure during module execution.", 
    "stdout": ""
}
```

I think that we're missing the code that takes the nspawn_args out of ansible_nspawn_extra_args and places it into play_context.nspawn_args.  I also saw that there's some nspawn_args in the config file (which are also not being used play_context).  I think maybe it should be made clear that those are defaults that the inventory variable can override?

dacresni commented:
Thank you all for working on this!

bcoca commented:
never got chance to fix the configs the way i wanted so added this #21596  for now

uSpike commented:
Hello, just wondering what the status of this feature is?  I'd love to use this connection driver but it has been removed.  Thanks.

tszym commented:
Hi,
I'v been quite busy lately and didn't took time to continue on this.
I'd like to work on it this week end.
